core-js/src/definitions.ts

```diff
@@ -0,0 +1,1421 @@
+    this.removeTypeReference(type as any);
+  }
+
+  protected abstract removeTypeReference(type: NamedType): void;
+
+  protected checkRemoval() {
+    if (this.isElementBuiltIn() && !Schema.prototype['canModifyBuiltIn'].call(this.schema()!)) {
+      throw buildError(`Cannot modify built-in ${this}`);
+    }
+    // We allow removals even on detached element because that doesn't particularly create issues (and we happen to do such
+    // removals on detached internally; though of course we could refactor the code if we wanted).
+  }
+
+  protected checkUpdate(addedElement?: { schema(): Schema | undefined }) {
+    if (this.isElementBuiltIn() && !Schema.prototype['canModifyBuiltIn'].call(this.schema()!)) {
+      throw buildError(`Cannot modify built-in ${this}`);
+    }
+    // Allowing to add element to a detached element would get hairy. Because that would mean that when you do attach an element,
+    // you have to recurse within that element to all children elements to check whether they are attached or not and to which
+    // schema. And if they aren't attached, attaching them as side-effect could be surprising (think that adding a single field
+    // to a schema could bring a whole hierachy of types and directives for instance). If they are attached, it only work if
+    // it's to the same schema, but you have to check.
+    // Overall, it's simpler to force attaching elements before you add other elements to them.
+    if (!this.schema()) {
+      throw buildError(`Cannot modify detached element ${this}`);
+    }
+    if (addedElement) {
+      const thatSchema = addedElement.schema();
+      if (thatSchema && thatSchema != this.schema()) {
+        throw buildError(`Cannot add element ${addedElement} to ${this} as its is attached another schema`);
+      }
+    }
+  }
+
+  abstract remove(): Referencer[];
+}
+
+abstract class BaseNamedElement<P extends SchemaElement<any, any> | Schema, Referencer> extends SchemaElement<P, Referencer> implements Named {
+  constructor(readonly name: string) {
+    super();
+  }
+}
+
+abstract class BaseNamedType<Referencer> extends BaseNamedElement<Schema, Referencer> {
+  protected readonly _referencers: Set<Referencer> = new Set();
+
+  constructor(name: string, readonly isBuiltIn: boolean = false) {
+    super(name);
+  }
+
+  private addReferencer(referencer: Referencer) {
+    this._referencers.add(referencer);
+  }
+
+  private removeReferencer(referencer: Referencer) {
+    this._referencers.delete(referencer);
+  }
+
+  get coordinate(): string {
+    return this.name;
+  }
+
+  *allChildrenElements(): Generator<SchemaElement<any, any>, void, undefined> {
+    // Overriden by those types that do have chidrens
+  }
+
+  protected isElementBuiltIn(): boolean {
+    return this.isBuiltIn;
+  }
+
+  /**
+   * Removes this type definition from its parent schema.
+   *
+   * After calling this method, this type will be "detached": it wil have no parent, schema, fields,
```

Comment:
I understand a detached type won't have a parent, but why would fields, values, directives, etc. be removed as well?

Reply:
There is a comment in `SchemaElement.checkUpdate()` that goes to try to explain the reasoning a bit. But long story short, having to deal with detached elements can be a bit messy.

If we only detach the parent, we may end up with a detached field whose type is still attached, and that sound messy/error prone. But more generally,  the problem is around behaviour when you "attach" (add) an detached element that itself has a deep hierarchy of elements. Think: I'm adding a field to a type, but that field link to another detached Object Type that itself may have fields linking to other detached types, implements detached interfaces and sprinkled in there there is some directe applications that indirectly link to "some" definition. First, we'd have to recursively explore that hierarchy to ensure nothing is attached somewhere else. But even then, we'd have to properly attach everything, which concretely means that the user might have call `addField` on a schema, and as a result 10 new types and whatnot are added to the schema on top of the field. Of course, maybe that's desired if you know what you're doing, but to me, this sounds like complex behaviour that is more likely to lead to surprising results.

Overall, the idea of the current patch is that while detached element are something we unfortunately have to deal, it tries to make it as innocuous/predictable as possible by ensure that a detached element is always "just an empty shell". Which is done effectively by:
1. no allowing any modification methods on a detached element (`SchemaElement.checkUpdate`).
2. cleaning up elements when they are removed.

Reply:
Yeah, dealing with detached elements is messy unfortunately. I agree we should avoid bringing in a whole hierarchy when re-attaching an element. I'm not sure the 'empty shell' semantics make sense to me though. My idea was that references to other elements would always be by name, so if you remove a field for example, you can attach it to a type in another schema but it would be your responsibility to ensure referenced elements are also added.